pkg/tracker/apitracker.go

```diff
@@ -0,0 +1,285 @@
+}
+
+// NewAPITracker creates a new instace of a API Tracker.
+func NewAPITracker(baseURL string) *APITracker {
+	return &APITracker{
+		baseURL:         baseURL,
+		httpClient:      *http.DefaultClient,
+		portStorage:     newPortStorage(),
+		ListenerTracker: NewListenerTracker(),
+	}
+}
+
+// Add a container ID and port mapping to the tracker and calls the
+// /services/forwarder/expose endpoint to forward the port mappings.
+func (a *APITracker) Add(containerID string, portMap nat.PortMap) error {
+	var errs []error
+
+	successfullyForwarded := make(nat.PortMap)
+
+	for portProto, portBindings := range portMap {
+		var tmpPortBinding []nat.PortBinding
+
+		for _, portBinding := range portBindings {
+			// The expose API only supports IPv4
+			ipv4, err := isIPv4(portBinding.HostIP)
+			if !ipv4 || err != nil {
+				continue
+			}
+
+			log.Debugf("calling %s API for the following port binding: %+v", exposeAPI, portBinding)
+
+			err = a.expose(
+				&types.ExposeRequest{
+					Local:  ipPortBuilder(portBinding.HostIP, portBinding.HostPort),
+					Remote: ipPortBuilder(hostSwitchIP, portBinding.HostPort),
+				})
+			if err != nil {
+				errs = append(errs, fmt.Errorf("exposing %+v failed: %w", portBinding, err))
+
+				continue
+			}
+
+			tmpPortBinding = append(tmpPortBinding, portBinding)
+		}
+
+		successfullyForwarded[portProto] = tmpPortBinding
+	}
+
+	a.portStorage.add(containerID, successfullyForwarded)
+
+	if len(errs) != 0 {
+		return fmt.Errorf("%w: %+v", ErrExposeAPI, errs)
+	}
+
+	return nil
+}
+
+// Get looks up the port mapping by containerID and returns the result.
+func (a *APITracker) Get(containerID string) nat.PortMap {
+	return a.portStorage.get(containerID)
+}
+
+// Remove a single entry from the port storage and calls the
+// /services/forwarder/unexpose endpoint to remove the forwarded the port mappings.
+func (a *APITracker) Remove(containerID string) error {
+	portMap := a.portStorage.get(containerID)
+	defer a.portStorage.remove(containerID)
+
+	var errs []error
+
+	for _, portBindings := range portMap {
+		for _, portBinding := range portBindings {
+			// The unexpose API only supports IPv4
+			ipv4, err := isIPv4(portBinding.HostIP)
+			if !ipv4 || err != nil {
+				continue
+			}
+
+			log.Debugf("calling %s API for the following port binding: %+v", unexposeAPI, portBinding)
+
+			err = a.unExpose(
+				&types.UnexposeRequest{
+					Local: ipPortBuilder(portBinding.HostIP, portBinding.HostPort),
+				})
+			if err != nil {
+				errs = append(errs,
+					fmt.Errorf("unexposing %+v failed: %w", portBinding, err))
+
+				continue
+			}
+		}
+	}
+
+	if len(errs) != 0 {
+		return fmt.Errorf("%w: %+v", ErrUnexposeAPI, errs)
+	}
+
+	return nil
+}
+
+// RemoveAll calls the /services/forwarder/unexpose
+// and removes all the port bindings from the tracker.
+func (a *APITracker) RemoveAll() error {
+	var errs []error
+
+	allPortMappings := a.portStorage.getAll()
+
+	for _, portMapping := range allPortMappings {
```

Comment:
We don't need the temporary `allPortMappings`, right?
```suggestion
	for _, portMapping := range a.portStorage.getAll() {
```